types/report-renderer.d.ts

```diff
@@ -0,0 +1,60 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+declare global {
+  module LH.Renderer {
+    export function renderFullReport(lhr: LH.Result, options?: ReportRendererOptions): HTMLElement;
+    export function renderReportComponents(
+      lhr: LH.Result,
+      options?: ReportRendererOptions
+    ): ReportComponents;
+
+    export function getFinalScreenshot(): void | string;
+
+    // Render gauge for a category
+    // category handles a bunch of plugin, n/a, and error cases. groupDefinitions only needed for PWA
+    export function renderGaugeForCategory(
+      category: LH.ReportCategory,
+      groupDefinitions?: Record<string, LH.Result.ReportGroup>
+    ): HTMLElement;
+
+    // Extra convience if you have just a category score.
+    export function renderGaugeForScore(num0to1: number): HTMLElement;
+  }
+
+  // TODO(paulirish): Refactor DOM to match. https://github.com/GoogleChrome/lighthouse/issues/12254#issuecomment-877520562
+  type ReportComponents = {
+    topbarEl: HTMLElement;
+    mainEl: HTMLElement;
+    headerEl: HTMLElement;
+    categoriesEl: HTMLElement;
+    footerEl: HTMLElement;
+  };
+
+  interface ReportRendererOptions {
+    /** DOM element that will the overlay DOM should be a child of.
+     * Between stacking contexts and z-index, the overlayParentEl should have a stacking/paint order high enough to cover all elements that the overlay should paint above.
+     * Defaults to the containerEl, but will be set in PSI to avoid being under the sticky header.
+     * @see https://philipwalton.com/articles/what-no-one-told-you-about-z-index/ */
+    overlayParentEl?: HTMLElement;
+
+    /** Callback running after a DOM element (like .lh-node or .lh-source-location) has been created */
+    onDetailsItemRendered?: (type: string, el: HTMLElement, value: any) => void;
+
+    /** Don't automatically apply dark-mode to dark based on (prefers-color-scheme: dark). (DevTools and PSINext don't want this.)
+     * Also, the fireworks easter-egg will want to flip to dark, so this setting will also disable chance of fireworks. */
+    disableAutoDarkModeAndFireworks?: boolean;
+
+    /** If defined, the 'Save as Gist' item in the topbar dropdown will be shown and when clicked, will run this function. */
+    onSaveGist?: (lhr: LH.Result) => string;
```

Comment:
currently there's a some indirection with how "save as gist" code runs ... in base ReportUIFeatures a `saveAsGist` fn simply throws and error, and Viewer overrides it https://github.com/GoogleChrome/lighthouse/blob/2d796ea/lighthouse-viewer/app/src/viewer-ui-features.js#L54 but that still defers to a passed-in callback function. the impl is `_onSaveJson` https://github.com/GoogleChrome/lighthouse/blob/2d796ea/lighthouse-viewer/app/src/lighthouse-report-viewer.js#L286

ideally the `report/renderer` code would just know how to save as gist and clients can enable as they wish, but we wouldn't want to ship firebase and idb to support `GithubApi` ...

perhaps a better approach would be providing a way to make a custom action in the dropdown menu. then in the options here, instead of a specific "give us a saveGist callback and we will wire it to an otherwise stubbed out and hidden dropdown action", we allow clients to define any action in that menu they want.

Reply:
>perhaps a better approach would be providing a way to make a custom action in the dropdown menu. then in the options here, instead of a specific "give us a saveGist callback and we will wire it to an otherwise stubbed out and hidden dropdown action", we allow clients to define any action in that menu they want.

I like this a lot, but I wonder if that opens us up to more flexibility (and complexity) than we need. The current implementation of dropdown actions calls `e.preventDefault` and manually handles each `data-action`. I could see an eventual `reportToolsMenuItems: Array<{label: string, iconUrl: string, onClick: () => void, position?: number}>` but without additional compelling use cases, I'm in favor of the `onSaveGist` approach here

Reply:
I agree that providing an API to define custom topbar menu actions would be ideal.. but i also agree with patrick that it's overkill for now. so i'd also like to keep this solution low key.